### PR TITLE
Task Dependency

### PR DESCRIPTION
**Pull request by callendorph**
This PR adds a new type of dependency, a `TaskDependency`. The idea is that a task dependency allows the user to define an entry point to a script or binary that `slm` will run during the dependency resolution process. This allows the user to construct some non-trivial dependency from whole cloth if necessary. 

## Example

I've setup my [lbstanza-z3](https://github.com/callendorph/lbstanza-z3) repo with slm a task dependency to build the Z3 shared library with [conan](https://conan.io/). Here is an example listing: 

```
$>  SLM_DEBUG=1 slm build z3-tests
slm: debug: build: Parsing git version: git version 2.39.2 (Apple Git-143)
slm: debug: build: Target: 'z3-tests'
slm: debug: build: Parsing stanza version: 0.18.58
slm: debug: build: Dep: libz3 Parent: lbstanza-z3 Task: One({ command = "./build_conan.sh" })

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=gnu17
compiler.libcxx=libc++
compiler.version=14
os=Macos
[options]
z3/*:shared=False

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=gnu17
compiler.libcxx=libc++
compiler.version=14
os=Macos


======== Computing dependency graph ========
Graph root
    conanfile.py (stanza-z3/None): /Users/callendorph/src/lbstanza-z3/conanfile.py
Requirements
    z3/4.12.2#e0d92a9719f08b1da98dacb71d7a2df0 - Cache

======== Computing necessary packages ========
Requirements
    z3/4.12.2#e0d92a9719f08b1da98dacb71d7a2df0:a77ad5ec6cec5767a19492a0e5aa905294be5886#5f4d531b34d8e1310c61020ecc2aacb6 - Cache

======== Installing packages ========
z3/4.12.2: Already installed! (1 of 1)
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.filenames' used in: z3/4.12.2
WARN: deprecated:     'cpp_info.names' used in: z3/4.12.2

======== Finalizing install (deploy, generators) ========
conanfile.py (stanza-z3/None): Conan built-in full deployer to /Users/callendorph/src/lbstanza-z3/build/static
conanfile.py (stanza-z3/None): Writing generators to /Users/callendorph/src/lbstanza-z3/build/static
conanfile.py (stanza-z3/None): Generator 'CMakeDeps' calling 'generate()'
conanfile.py (stanza-z3/None): Generator 'CMakeToolchain' calling 'generate()'
conanfile.py (stanza-z3/None): CMakeToolchain generated: conan_toolchain.cmake
conanfile.py (stanza-z3/None): Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.py (stanza-z3/None): If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/Users/callendorph/src/lbstanza-z3/build/static/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.py (stanza-z3/None): CMakeToolchain generated: CMakePresets.json
conanfile.py (stanza-z3/None): Generating aggregated env files
conanfile.py (stanza-z3/None): Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
Install finished successfully
slm: debug: build: /usr/local/bin/conan
LibZ3: Building Static Library
./build/static/full_deploy/host/z3/4.12.2/Release/x86_64/lib/libz3.a -> ./build/content/libz3.a
LibZ3: Build Complete
slm: debug: build: done resolving 'lbstanza-z3'
slm: debug: build: done parsing dependency graph
slm: debug: build: Project: Stanza Version Spec: None
slm: debug: build: Project: Compatible: true | No Stanza Version Requirement Specified
slm: debug: build: with stanza='stanza'
slm: debug: build: with build args ''
slm: debug: build: Build Version: 0.1.0
slm: debug: build: Stanza: stanza, build, z3-tests
Build target z3-tests is already up-to-date.
``` 

With this new task mechanism, the user can now run: 

```
$> slm add -git "callendorph/lbstanza-z3"
```

and `slm` manages all of the dependencies and setup.

## Details 

I've added a new sub-command `add-task` to allow the user to create new task dependencies. Task dependencies can be removed using the standard `remove` command. 

Task dependencies look like a normal dependency in the `slm.toml` file with some new features:

```
[dependencies]
libz3.task = { command = "./build_conan.sh" }
libz3.task.windows = { command = "./build_conan.ps1" }
```

Here you can see I define one default task `libz3.task` that would run on all platforms. I then add a subsequent definition that targets, explicitly, the `windows` platform. This second definition overrides the first when running on windows. 

You can additionally pass the following arguments: 

1.  `args` - normal command line arguments as a string
2.  `working-dir` - Working directory to run the command. By default it runs in the project root directory where the `slm.toml` file is found.
3.  `env-vars` - Define additional environment variables as necessary. 

## Limitations

I don't think this current implementation will work with `apt`. For example, if you tried: 

 ```
[dependencies]
libpng.task.linux = { command = "sudo" args="apt install libpng-dev" }
```

I think it would fail because of interactivity required to get the users password. But if the user was setup with no password required for sudo - it might work fine.

### CODE REVIEW

File: src/commands/add-task.stanza [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  match(platform?):
+    (x:None):
+      if task?(existing) is-not None and not force:
+        throw $ ValueError("Task '%_' Already Exists" % [name])
+
+      sub-task?(existing, None())
+    (x:One<String>):
+      val platform = value(x)
+      val plat-tasks = to-hashtable<String, Task> $ platform-tasks(existing)
+      val plat-task = to-maybe $ get?{_, platform, false} $ plat-tasks
```
Comment (tjknoth) [inline src/commands/add-task.stanza:81]: Can we remove the task more directly? Isn't `platform-tasks(existing)` already a `HashTable` that you can mutate?

File: src/dependencies.stanza [hunk 2]
```diff
@@ -5,38 +5,62 @@ defpackage slm/dependencies:
...
   import semver
   import term-colors
-  import toml/file
-  import toml/parser
-  import toml/table
+  import toml
 
   import slm/flags
-  import slm/dependency
+  import slm/dep-types
```
Comment (tjknoth) [inline src/dependencies.stanza:11]: thought you implemented dependent types for a sec....

**Review by tjknoth — APPROVED** [review]
Looks pretty robust!
For the record this should work fine with `apt` (when you can `sudo`) -- there are flags to automatically answer "yes" to all interactive polls.
One question I probably could have answered if I read the code more closely: how do we know if `TaskDependency`s are up-to-date? just a hash of the shell command?

File: src/utils.stanza [hunk 3]
```diff
@@ -81,4 +82,19 @@ public defn split-any (s:String, splitters:String) -> Tuple<String> :
...
+Check that the pass string is non-empty
+
+This is intended to be used with the `ensure` functionality of
+`defstruct` member construction.
+<DOC>
+public defn ensure-not-empty! (field:String, v:String):
+  if length(v) == 0:
+    throw $ ValueError("Field[%_] - Expected Non-empty String - Received: %~" % [field, v])
+
+public defn ensure-not-empty! (field:String, v?:Maybe<String>):
```
[the remark this answers is not in the document]
Comment (callendorph) [reply src/utils.stanza:97]: Ya - I don't know. Its 6 one way, half-dozen another. Frankly, I end up doing a lot of the same things over and over again. Some of the `ensure` stuff should be either put into a library or put into core or something.

File: src/commands/add-task.stanza (hunk 1 above)
Comment (callendorph) [reply src/commands/add-task.stanza:81]: Ya - I had originally implemented as a `Tuple<KeyValue>` and so I couldn't remove it directly. I think I didn't update after making the change to `HashTable`. I think I was also trying to keep it as immutable as possible but I'm not sure that really matters in this case.

File: src/dependencies.stanza (hunk 2 above)
Comment (callendorph) [reply src/dependencies.stanza:11]: Ha - I'm not that ambitious.

**Comment by callendorph** [comment]
> how do we know if TaskDependencys are up-to-date? just a hash of the shell command?

My model is that the `TaskDependency` is idempotent. So we just run the script on every dependency resolution process. There is no "up to date" because I don't know exactly what up to date will mean. 

Maybe this isn't a good idea, but it is a pretty simple/straightforward approach. It seems to work well with `conan2` because conan caches the results of previous builds. I find with `lbstanza-z3`- I can integrate the script and it does not appreciably change the build time after the build has completed at least once.

**Comment by tjknoth** [comment]
> > how do we know if TaskDependencys are up-to-date? just a hash of the shell command?
> 
> My model is that the `TaskDependency` is idempotent. So we just run the script on every dependency resolution process. There is no "up to date" because I don't know exactly what up to date will mean.
> 
> Maybe this isn't a good idea, but it is a pretty simple/straightforward approach. It seems to work well with `conan2` because conan caches the results of previous builds. I find with `lbstanza-z3`- I can integrate the script and it does not appreciably change the build time after the build has completed at least once.

Seems like a pragmatic and reasonable assumption, makes sense to me!

**Comment by callendorph** [comment]
OK - I fixed the complexity in the remove func you pointed out. 
I added some additional documentation about idempotency requirement.

**Pull request merged**